### PR TITLE
raftstore: handle apply concurrently

### PR DESCRIPTION
**Pull request by BusyJay**
@ngaut @siddontang @disksing @zhangjinpeng1987 PTAL

### CODE REVIEW

File: src/raftstore/store/store.rs [hunk 1]
```diff
@@ -483,26 +488,48 @@ impl<T: Transport, C: PdClient> Store<T, C> {
...
+        let (tx, rx) = mpsc::channel();
+        for &region_id in &ids {
+            if let Some(mut peer) = self.region_peers.remove(&region_id) {
+                let sender = tx.clone();
+                let trans = self.trans.clone();
+                self.apply_pool.execute(move || {
+                    let res = peer.handle_raft_ready(&trans);
+                    if let Err(SendError((peer, _))) = sender.send((peer, res)) {
+                        // TODO: retry here.
+                        error!("failed to report ready result of [{}] {}, will retry",
```
Comment (siddontang) [inline src/raftstore/store/store.rs]: I think panic is ok here.

File: src/raftstore/store/store.rs [hunk 2]
```diff
@@ -483,26 +488,48 @@ impl<T: Transport, C: PdClient> Store<T, C> {
...
+            let (peer, res) = rx.recv().expect("recv ready result can't failed");
+
+            let region_id = peer.region_id();
+            self.region_peers.insert(region_id, peer);
+
+            match res {
+                Err(e) => {
+                    // TODO: should we panic or shutdown the store?
+                    error!("handle raft ready at region {} err: {:?}", region_id, e);
+                    has_err = true;
```
Comment (siddontang) [inline src/raftstore/store/store.rs]: maybe no need using has_err here, origin outer just log this error too.

File: src/raftstore/store/store.rs (hunk 1 above)
Comment (BusyJay) [inline src/raftstore/store/store.rs]: No need panic here, receiver will panic in the end.

File: src/raftstore/store/store.rs (hunk 2 above)
Comment (BusyJay) [inline src/raftstore/store/store.rs]: I think it's better to propagate the error.

File: src/raftstore/store/store.rs [hunk 3]
```diff
@@ -511,7 +538,11 @@ impl<T: Transport, C: PdClient> Store<T, C> {
                   pending_count,
                   t.elapsed());
 
-        Ok(())
+        if has_err {
+            Err(box_err!("some error happen during handle ready."))
```
Comment (siddontang) [inline src/raftstore/store/store.rs]: errors

File: src/raftstore/store/store.rs [hunk 4]
```diff
@@ -483,26 +488,48 @@ impl<T: Transport, C: PdClient> Store<T, C> {
...
-            if let Some(ready_result) = ready_result {
-                if let Err(e) = self.on_ready_result(region_id, ready_result) {
-                    error!("handle raft ready result at region {} err: {:?}",
-                           region_id,
-                           e);
-                    return Err(e);
+        let mut has_err = false;
+        for _ in 0..send_cnt {
+            // if err, means we lost some peer.
+            let (peer, res) = rx.recv().expect("recv ready result can't failed");
```
Comment (siddontang) [inline src/raftstore/store/store.rs]: can't fail

**Comment by siddontang** [comment]
Rest LGTM

File: src/raftstore/store/store.rs [hunk 5]
```diff
@@ -483,26 +488,48 @@ impl<T: Transport, C: PdClient> Store<T, C> {
...
-            if let Some(ready_result) = ready_result {
-                if let Err(e) = self.on_ready_result(region_id, ready_result) {
-                    error!("handle raft ready result at region {} err: {:?}",
-                           region_id,
-                           e);
-                    return Err(e);
+        let mut has_err = false;
+        for _ in 0..send_cnt {
+            // if err, means we lost some peer.
+            let (peer, res) = rx.recv().expect("recv ready result can't fail");
```
Comment (ngaut) [inline src/raftstore/store/store.rs]: s/can't/shouldn't/

File: src/raftstore/store/store.rs [hunk 6]
```diff
@@ -483,26 +488,48 @@ impl<T: Transport, C: PdClient> Store<T, C> {
...
+
```
Comment (ngaut) [inline src/raftstore/store/store.rs]: Remove blank line.

File: src/raftstore/store/store.rs [hunk 7]
```diff
@@ -483,26 +488,48 @@ impl<T: Transport, C: PdClient> Store<T, C> {
...
+        }
 
-            if let Some(ready_result) = ready_result {
-                if let Err(e) = self.on_ready_result(region_id, ready_result) {
-                    error!("handle raft ready result at region {} err: {:?}",
-                           region_id,
-                           e);
-                    return Err(e);
+        let mut has_err = false;
+        for _ in 0..send_cnt {
```
Comment (zhangjinpeng1987) [inline src/raftstore/store/store.rs]: Is it possible the raft thread dont need wait?

Comment (ngaut) [inline src/raftstore/store/store.rs]: It's possible.

Comment (BusyJay) [inline src/raftstore/store/store.rs]: No, raft is not thread safe. So must wait before moving on.

**Comment by ngaut** [comment]
LGTM

**Comment by BusyJay** [comment]
PTAL

File: src/raftstore/store/store.rs [hunk 8]
```diff
@@ -483,26 +488,57 @@ impl<T: Transport, C: PdClient> Store<T, C> {
...
-                        // TODO: should we panic or shutdown the store?
-                        error!("handle raft ready at region {} err: {:?}", region_id, e);
-                        return Err(e);
+        let mut sent_cnt = 0;
+        let (tx, rx) = mpsc::channel();
+        for &region_id in &ids {
+            if let Some(mut peer) = self.region_peers.remove(&region_id) {
+                let sender = tx.clone();
+                let trans = self.trans.clone();
+                self.apply_pool.execute(move || {
```
Comment (siddontang) [inline src/raftstore/store/store.rs]: s/apply_pool/ready_pool/

Comment (BusyJay) [inline src/raftstore/store/store.rs]: `apply_pool` means the thread is used to do the apply jobs. `ready_pool` sounds like the pool is used to store ready.

Comment (siddontang) [inline src/raftstore/store/store.rs]: apply seems only to apply raft logs, but here we need to handle other raft steps too.

File: src/raftstore/store/store.rs [hunk 9]
```diff
@@ -483,26 +488,57 @@ impl<T: Transport, C: PdClient> Store<T, C> {
...
-                match peer.handle_raft_ready(&self.trans) {
-                    Err(e) => {
-                        // TODO: should we panic or shutdown the store?
-                        error!("handle raft ready at region {} err: {:?}", region_id, e);
-                        return Err(e);
+        let mut sent_cnt = 0;
+        let (tx, rx) = mpsc::channel();
+        for &region_id in &ids {
+            if let Some(mut peer) = self.region_peers.remove(&region_id) {
+                let sender = tx.clone();
```
Comment (siddontang) [inline src/raftstore/store/store.rs]: if the peer raft is not ready, we can skip it directly.

File: src/raftstore/store/peer.rs [hunk 10]
```diff
@@ -358,6 +360,29 @@ impl Peer {
...
+        // For more details, check raft thesis 10.2.1
+        if self.is_leader() {
+            try!(self.send(trans, &mut ready.messages));
+        }
+
+        // Can we just ignore hardstate here?
+        if ready.entries.is_empty() && ready.hs.is_none() &&
+           (self.get_store().rl().is_applying_snap() ||
+            ready.committed_entries.is_empty() && raft::is_empty_snap(&ready.snapshot)) {
+            try!(self.send(trans, &mut ready.messages));
```
Comment (siddontang) [inline src/raftstore/store/peer.rs]: if the peer is leader, we may send the message again?

Comment (BusyJay) [inline src/raftstore/store/peer.rs]: No, `ready.messages` will be drained when sent.

**Comment by BusyJay** [comment]
Performance seems not being improved, close now.

**Pull request closed without merging**